docs/module-caching.md

```diff
@@ -0,0 +1,181 @@
+# Modules evaluation and caching
+
+This proposal aims to enforce some guarantees about how many times modules are evaluated and in which contexts. ECMAScript already guarantees that importing the same module from the same file results in a single evaluation, but Module Blocks increase the complexity because they can be passed around.
+
+## Invariants
+
+1. Importing a module with the same string specifier twice from the same file results in a single evaluation (this is already guaranteed by ecma262):
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   import { check as c2 } from "./file.js";
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   import { check as c1 } from "./file.js";
+   const { check: c2 } = await import("./file.js");
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+   ```js
+   // main.js
+   const { check: c1 } = await import("./file.js");
+   const { check: c2 } = await import("./file.js");
+   assert(c1 === c2);
+   
+   // file.js
+   export const check = {};
+   ```
+2. Importing a module with the same string specifier from two different modules defined in the same file and evaluated the same Realm results in a single evaluation:
```

Comment:
```suggestion
2. Importing a module with the same string specifier from the top-level or from any module blocks defined in the same source text module and evaluated the same Realm results in a single evaluation:
```

Perhaps for even further clarify we should include an example of a module block nested inside of another module block?